File: backend/pipeline_orchestrator/src/api/concepts.py
"""
API endpoints for concept summary generation
"""

from fastapi import APIRouter, HTTPException, Depends
from typing import List, Optional
from uuid import UUID

from ..models.concepts import (
    ConceptSummaryRequest,
    ConceptSummaryResponse,
    ConceptCreate,
    Concept,
    ConceptUpdate,
    ConceptWithSummaries
)
from ..services.concept_summary_service import ConceptSummaryService
from ..llm.claude_client import ClaudeClient

router = APIRouter(prefix="/api/concepts", tags=["concepts"])


def get_concept_service() -> ConceptSummaryService:
    """Dependency to get concept summary service"""
    return ConceptSummaryService(ClaudeClient())


@router.post("/generate-summary", response_model=ConceptSummaryResponse)
async def generate_concept_summary(
    request: ConceptSummaryRequest,
    service: ConceptSummaryService = Depends(get_concept_service)
) -> ConceptSummaryResponse:
    """
    Generate a one-line summary for a concept using AI

    Args:
        request: Concept details including name, grade level, and context

    Returns:
        Generated summary with alternatives and confidence score
    """
    try:
        summary = await service.generate_summary(request)
        return summary
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error generating summary: {str(e)}")


@router.post("/generate-batch", response_model=dict)
async def generate_batch_summaries(
    concepts: List[str],
    grade_level: int,
    module_context: Optional[str] = None,
    service: ConceptSummaryService = Depends(get_concept_service)
) -> dict:
    """
    Generate summaries for multiple concepts at once

    Args:
        concepts: List of concept names
        grade_level: Target grade level
        module_context: Optional context about the module

    Returns:
        Dictionary mapping concept names to their summaries
    """
    try:
        summaries = await service.generate_summaries_for_concepts(
            concepts=concepts,
            grade_level=grade_level,
            module_context=module_context
        )
        return summaries
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error generating batch summaries: {str(e)}")


@router.post("/validate-summary")
async def validate_concept_summary(
    concept_name: str,
    summary: str,
    grade_level: int,
    service: ConceptSummaryService = Depends(get_concept_service)
):
    """
    Validate a concept summary for quality and appropriateness

    Args:
        concept_name: Name of the concept
        summary: The summary to validate
        grade_level: Target grade level

    Returns:
        Validation results with scores and feedback
    """
    try:
        validation = await service.validate_summary(
            concept_name=concept_name,
            summary=summary,
            grade_level=grade_level
        )
        return validation
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error validating summary: {str(e)}")


@router.post("/regenerate-summary", response_model=ConceptSummaryResponse)
async def regenerate_concept_summary(
    concept_name: str,
    grade_level: int,
    feedback: Optional[str] = None,
    previous_summary: Optional[str] = None,
    service: ConceptSummaryService = Depends(get_concept_service)
) -> ConceptSummaryResponse:
    """
    Regenerate a concept summary with teacher feedback

    Args:
        concept_name: Name of the concept
        grade_level: Target grade level
        feedback: Teacher feedback on previous summary
        previous_summary: Previous summary to improve upon

    Returns:
        New generated summary
    """
    try:
        summary = await service.regenerate_summary(
            concept_name=concept_name,
            grade_level=grade_level,
            feedback=feedback,
            previous_summary=previous_summary
        )
        return summary
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error regenerating summary: {str(e)}")


@router.get("/health")
async def health_check():
    """Health check endpoint"""
    return {"status": "healthy", "service": "concept-summary-service"}

File: backend/api_gateway/src/routes/concepts.js
import express from 'express';
import axios from 'axios';

const router = express.Router();

const ORCHESTRATOR_URL = process.env.ORCHESTRATOR_URL || 'http://localhost:8000';

/**
 * Generate concept summary using AI
 * POST /api/concepts/generate-summary
 */
router.post('/generate-summary', async (req, res) => {
  try {
    const { concept_name, grade_level, concept_description, module_context, related_concepts } = req.body;

    if (!concept_name || !grade_level) {
      return res.status(400).json({
        error: 'concept_name and grade_level are required'
      });
    }

    // Forward request to Python orchestrator
    const response = await axios.post(
      `${ORCHESTRATOR_URL}/api/concepts/generate-summary`,
      {
        concept_name,
        grade_level,
        concept_description,
        module_context,
        related_concepts: related_concepts || []
      },
      {
        timeout: 30000 // 30 second timeout for AI generation
      }
    );

    res.json(response.data);
  } catch (error) {
    console.error('Error generating concept summary:', error.message);
    res.status(error.response?.status || 500).json({
      error: error.response?.data?.detail || 'Failed to generate concept summary'
    });
  }
});

/**
 * Generate summaries for multiple concepts
 * POST /api/concepts/generate-batch
 */
router.post('/generate-batch', async (req, res) => {
  try {
    const { concepts, grade_level, module_context } = req.body;

    if (!concepts || !Array.isArray(concepts) || concepts.length === 0) {
      return res.status(400).json({
        error: 'concepts array is required and must not be empty'
      });
    }

    if (!grade_level) {
      return res.status(400).json({
        error: 'grade_level is required'
      });
    }

    // Forward request to Python orchestrator
    const response = await axios.post(
      `${ORCHESTRATOR_URL}/api/concepts/generate-batch`,
      null,
      {
        params: {
          grade_level,
          module_context
        },
        data: concepts,
        timeout: 60000 // 60 second timeout for batch generation
      }
    );

    res.json(response.data);
  } catch (error) {
    console.error('Error generating batch summaries:', error.message);
    res.status(error.response?.status || 500).json({
      error: error.response?.data?.detail || 'Failed to generate batch summaries'
    });
  }
});

/**
 * Validate a concept summary
 * POST /api/concepts/validate-summary
 */
router.post('/validate-summary', async (req, res) => {
  try {
    const { concept_name, summary, grade_level } = req.body;

    if (!concept_name || !summary || !grade_level) {
      return res.status(400).json({
        error: 'concept_name, summary, and grade_level are required'
      });
    }

    // Forward request to Python orchestrator
    const response = await axios.post(
      `${ORCHESTRATOR_URL}/api/concepts/validate-summary`,
      null,
      {
        params: {
          concept_name,
          summary,
          grade_level
        },
        timeout: 20000
      }
    );

    res.json(response.data);
  } catch (error) {
    console.error('Error validating summary:', error.message);
    res.status(error.response?.status || 500).json({
      error: error.response?.data?.detail || 'Failed to validate summary'
    });
  }
});

/**
 * Regenerate concept summary with feedback
 * POST /api/concepts/regenerate-summary
 */
router.post('/regenerate-summary', async (req, res) => {
  try {
    const { concept_name, grade_level, feedback, previous_summary } = req.body;

    if (!concept_name || !grade_level) {
      return res.status(400).json({
        error: 'concept_name and grade_level are required'
      });
    }

    // Forward request to Python orchestrator
    const response = await axios.post(
      `${ORCHESTRATOR_URL}/api/concepts/regenerate-summary`,
      null,
      {
        params: {
          concept_name,
          grade_level,
          feedback,
          previous_summary
        },
        timeout: 30000
      }
    );

    res.json(response.data);
  } catch (error) {
    console.error('Error regenerating summary:', error.message);
    res.status(error.response?.status || 500).json({
      error: error.response?.data?.detail || 'Failed to regenerate summary'
    });
  }
});

/**
 * Health check for concepts service
 * GET /api/concepts/health
 */
router.get('/health', async (req, res) => {
  try {
    const response = await axios.get(`${ORCHESTRATOR_URL}/api/concepts/health`, {
      timeout: 5000
    });
    res.json(response.data);
  } catch (error) {
    res.status(500).json({
      status: 'unhealthy',
      error: error.message
    });
  }
});

export default router;
